lib/API/VK/Device.cpp

```diff
@@ -1346,3 +1272,136 @@ llvm::Error Device::initializeVKDevices(const DeviceConfig Config) {
 }
 
 void Device::cleanupVKDevices() { VKContext::instance().cleanup(); }
+
+class VKValidationError : public llvm::ErrorInfo<VKValidationError> {
+public:
+  static char ID;
+
+  VKValidationError(VkDebugUtilsMessageSeverityFlagBitsEXT Severity,
+                    const VkDebugUtilsMessengerCallbackDataEXT *Data)
+      : MessageSeverity(Severity) {
+    MessageID = Data->pMessageIdName;
+    Message = Data->pMessage;
+    for (uint32_t I = 0; I < Data->objectCount; I++)
+      if (Data->pObjects[I].pObjectName)
+        Objects.emplace_back(std::string(Data->pObjects[I].pObjectName));
+  }
+
+  void log(llvm::raw_ostream &OS) const override {
+    OS << "Validation " << getMessageSeverityString(MessageSeverity);
+    OS << ": [ " << MessageID << " ]\n";
+    OS << Message;
+
+    for (const auto &S : Objects)
+      OS << "\n[" << S << "]";
+    OS << '\n';
+  }
+
+  std::error_code convertToErrorCode() const override {
+    return llvm::inconvertibleErrorCode();
+  }
+
+private:
+  VkDebugUtilsMessageSeverityFlagBitsEXT MessageSeverity;
+  std::string Message;
+  std::string MessageID;
+  std::vector<std::string> Objects;
```

Comment:
SmallVector would be slightly more efficient here (it doesn't really matter though)